File: scala/src/test/scala/tennis/TennisTest.scala
package tennis

import org.junit.Assert._
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized
import org.junit.runners.Parameterized.Parameters
import tennis.TennisTest.{player1Name, player2Name}

import java.util
import scala.util.Random

@RunWith(classOf[Parameterized])
class TennisTest(params: TennisTestCase) {

  def checkScores(game: TennisGame) {
    val highestScore = Math.max(params.player1Score, params.player2Score)
      for (i <- 0 until highestScore by 1) {
          if (i < params.player1Score)
              game.wonPoint(player1Name)
          if (i < params.player2Score)
              game.wonPoint(player2Name)
      }
      assertEquals(params.expectedScore, game.calculateScore())
  }

  @Test
  def checkGame1() {
      val game = new TennisGame1(player1Name, player2Name)
      checkScores(game)
  }

  @Test
  def checkGame2() {
      val game = new TennisGame2(player1Name, player2Name)
      checkScores(game)
  }


  @Test
  def checkGame3() {
      val game = new TennisGame3(player1Name, player2Name)
      checkScores(game)
  }

}

object TennisTest {
  val player1Name = "player1-" + aRandomString
  val player2Name = "player2-" + aRandomString
  
  @Parameters(name = "{0}")
  def getAllScores() : java.util.Collection[Array[TennisTestCase]] = {
      var list = new util.ArrayList[Array[TennisTestCase]]();

      val args = Array(
        new TennisTestCase( 0, 0, "Love-All"),
        new TennisTestCase( 1, 1, "Fifteen-All" ),
        new TennisTestCase( 2, 2, "Thirty-All"),
        new TennisTestCase( 3, 3, "Deuce"),
        new TennisTestCase( 4, 4, "Deuce"),

        new TennisTestCase( 1, 0, "Fifteen-Love"),
        new TennisTestCase( 0, 1, "Love-Fifteen"),
        new TennisTestCase( 2, 0, "Thirty-Love"),
        new TennisTestCase( 0, 2, "Love-Thirty"),
        new TennisTestCase( 3, 0, "Forty-Love"),
        new TennisTestCase( 0, 3, "Love-Forty"),
        new TennisTestCase( 4, 0, s"Win for $player1Name"),
        new TennisTestCase( 0, 4, s"Win for $player2Name"),

        new TennisTestCase( 2, 1, "Thirty-Fifteen"),
        new TennisTestCase( 1, 2, "Fifteen-Thirty"),
        new TennisTestCase( 3, 1, "Forty-Fifteen"),
        new TennisTestCase( 1, 3, "Fifteen-Forty"),
        new TennisTestCase( 4, 1, s"Win for $player1Name"),
        new TennisTestCase( 1, 4, s"Win for $player2Name"),

        new TennisTestCase( 3, 2, "Forty-Thirty"),
        new TennisTestCase( 2, 3, "Thirty-Forty"),
        new TennisTestCase( 4, 2, s"Win for $player1Name"),
        new TennisTestCase( 2, 4, s"Win for $player2Name"),

        new TennisTestCase( 4, 3, s"Advantage $player1Name"),
        new TennisTestCase( 3, 4, s"Advantage $player2Name"),
        new TennisTestCase( 5, 4, s"Advantage $player1Name"),
        new TennisTestCase( 4, 5, s"Advantage $player2Name"),
        new TennisTestCase( 15, 14, s"Advantage $player1Name"),
        new TennisTestCase( 14, 15, s"Advantage $player2Name"),

        new TennisTestCase( 6, 4, s"Win for $player1Name"),
        new TennisTestCase( 4, 6, s"Win for $player2Name"),
        new TennisTestCase( 16, 14, s"Win for $player1Name"),
        new TennisTestCase( 14, 16, s"Win for $player2Name")
     )
     args.foreach(n => list.add(Array(n)))
     return list
  }

  private def aRandomString = {
    Random.nextInt(1000).toString
  }
}

File: java/src/test/java/TennisTest.java
import org.junit.jupiter.params.ParameterizedTest;

import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.assertEquals;
import org.junit.jupiter.params.provider.MethodSource;

public class TennisTest {

    public static Stream<Object[]> getAllScores() {
        return Stream.of(new Object[][]{
                {0, 0, "Love-All"},
                {1, 1, "Fifteen-All"},
                {2, 2, "Thirty-All"},
                {3, 3, "Deuce"},
                {4, 4, "Deuce"},

                {1, 0, "Fifteen-Love"},
                {0, 1, "Love-Fifteen"},
                {2, 0, "Thirty-Love"},
                {0, 2, "Love-Thirty"},
                {3, 0, "Forty-Love"},
                {0, 3, "Love-Forty"},
                {4, 0, "Win for player1"},
                {0, 4, "Win for player2"},

                {2, 1, "Thirty-Fifteen"},
                {1, 2, "Fifteen-Thirty"},
                {3, 1, "Forty-Fifteen"},
                {1, 3, "Fifteen-Forty"},
                {4, 1, "Win for player1"},
                {1, 4, "Win for player2"},

                {3, 2, "Forty-Thirty"},
                {2, 3, "Thirty-Forty"},
                {4, 2, "Win for player1"},
                {2, 4, "Win for player2"},

                {4, 3, "Advantage player1"},
                {3, 4, "Advantage player2"},
                {5, 4, "Advantage player1"},
                {4, 5, "Advantage player2"},
                {15, 14, "Advantage player1"},
                {14, 15, "Advantage player2"},

                {6, 4, "Win for player1"},
                {4, 6, "Win for player2"},
                {16, 14, "Win for player1"},
                {14, 16, "Win for player2"},
        });
    }

    private static void checkAllScores(int player1Points, int player2Points, String expectedScore, TennisGame game) {
        int highestScore = Math.max(player1Points, player2Points);
        for (int i = 0; i < highestScore; i++) {
            if (i < player1Points)
                game.wonPoint("player1");
            if (i < player2Points)
                game.wonPoint("player2");
        }
        assertEquals(expectedScore, game.getScore());
    }

    @ParameterizedTest
    @MethodSource("getAllScores")
    public void checkAllScoresTennisGame1(int player1Points, int player2Points, String expectedScore) {
        TennisGame game = new TennisGame1("player1", "player2");
        checkAllScores(player1Points, player2Points, expectedScore, game);
    }

    @ParameterizedTest
    @MethodSource("getAllScores")
    public void checkAllScoresTennisGame2(int player1Points, int player2Points, String expectedScore) {
        TennisGame game = new TennisGame2("player1", "player2");
        checkAllScores(player1Points, player2Points, expectedScore, game);
    }

    @ParameterizedTest
    @MethodSource("getAllScores")
    public void checkAllScoresTennisGame3(int player1Points, int player2Points, String expectedScore) {
        TennisGame game = new TennisGame3("player1", "player2");
        checkAllScores(player1Points, player2Points, expectedScore, game);
    }

    @ParameterizedTest
    @MethodSource("getAllScores")
    public void checkAllScoresTennisGame4(int player1Points, int player2Points, String expectedScore) {
        TennisGame game = new TennisGame4("player1", "player2");
        checkAllScores(player1Points, player2Points, expectedScore, game);
    }

    @ParameterizedTest
    @MethodSource("getAllScores")
    public void checkAllScoresTennisGame5(int player1Points, int player2Points, String expectedScore) {
        TennisGame game = new TennisGame5("player1", "player2");
        checkAllScores(player1Points, player2Points, expectedScore, game);
    }

    @ParameterizedTest
    @MethodSource("getAllScores")
    public void checkAllScoresTennisGame6(int player1Points, int player2Points, String expectedScore) {
        TennisGame game = new TennisGame6("player1", "player2");
        checkAllScores(player1Points, player2Points, expectedScore, game);
    }


    @ParameterizedTest
    @MethodSource("getAllScores")
    public void checkAllScoresTennisGame7(int player1Points, int player2Points, String expectedScore) {
        TennisGame game = new TennisGame7("player1", "player2");
        checkAllScores(player1Points, player2Points, "Current score: " + expectedScore + ", enjoy your game!", game);
    }
}
